[BUG] Relative imports '.' and '..' do not work

### Bug Report Quick Checklist

- [x] I am on the latest version of Snowpack & all plugins.
- [x] I use package manager **yarn/npm** (Fill in: npm, yarn, pnpm, etc).
- [x] I run Snowpack on OS **mac** (Fill in: Windows, Mac, Linux, etc).
- [x] I run Snowpack on Node.js v12+

### Describe the bug

Extracted from https://github.com/snowpackjs/snowpack/pull/2707#issuecomment-790926970, so it doesn't get lost.

I've found an issue with relative paths imports - `import {BLAH} from '.'` and `import {BLAH} from '..'` do not work as expected, however `import {BLAH} from './'` and `import {BLAH} from '../'`do work correctly. 

It seems this was a problem in 3.0.x too, however 3.0.x compiled and served the file with incorrect content and the error occured at runtime. However in the 3.1.0-pre.11 the file with problematic code now returns a 500 error, with an unhelpful error message.

[Reproduction is available here](https://github.com/BPScott/snowpack-test/tree/main/bare-relative-import).

Given the input ([found in this file](https://github.com/BPScott/snowpack-test/blob/main/bare-relative-import/src/subfolder/subsubfolder/value.ts))

```js
import {value1} from '.';
import {value2} from '..';

import {value3} from './';
import {value4} from '../';
```

In 3.0.x it compiles to the incorrect output:

```js
import {value1} from '.';
import {value2} from '..';

import {value3} from './index.js';
import {value4} from '../index.js';
```

In 3.1.0-pre.11 it errors during compile:

```
Error: Package(s) failed to build: Cannot find module '..' from '/Users/ben/src/github.com/BPScott/snowpack-test/bare-relative-import/package.json'
```

### To Reproduce

[Reproduction is available here](https://github.com/BPScott/snowpack-test/tree/main/bare-relative-import).

Clone that repo, move into the `bare-relative-import` subfolder and run `npm install` and `npm start`

### Expected behavior

I would expect the code:

```js
import {value1} from '.';
import {value2} from '..';

import {value3} from './';
import {value4} from '../';
```

to compile to:

```js
import {value1} from './index.js';
import {value2} from '../index.js';

import {value3} from './index.js';
import {value4} from '../index.js';
```

--- a/snowpack/src/build/import-resolver.ts
+++ b/snowpack/src/build/import-resolver.ts
@@ -6,6 +6,7 @@ import {
   findMatchingAliasEntry,
   getExtensionMatch,
   hasExtension,
+  isPathImport,
   isRemoteUrl,
   replaceExtension,
 } from '../util';
@@ -30,7 +31,7 @@ function resolveSourceSpecifier(
 
   // Handle directory imports (ex: "./components" -> "./components/index.js")
   if (lazyFileStat && lazyFileStat.isDirectory()) {
-    const trailingSlash = lazyFileLoc.endsWith('/') ? '' : '/';
+    const trailingSlash = lazyFileLoc.endsWith(path.sep) ? '' : path.sep;
     lazyFileLoc = lazyFileLoc + trailingSlash + 'index.js';
   } else if (lazyFileStat && lazyFileStat.isFile()) {
     lazyFileLoc = lazyFileLoc;
@@ -99,11 +100,11 @@ export function createImportResolver({fileLoc, config}: {fileLoc: string; config
     if (config.packageOptions.external?.includes(spec)) {
       return spec;
     }
-    if (spec.startsWith('/')) {
+    if (spec[0] === '/') {
       return spec;
     }
-    if (spec.startsWith('./') || spec.startsWith('../') || spec === '.') {
-      const importedFileLoc = path.resolve(path.dirname(fileLoc), spec);
+    if (isPathImport(spec)) {
+      const importedFileLoc = path.resolve(path.dirname(fileLoc), path.normalize(spec));
       return resolveSourceSpecifier(importedFileLoc, {parentFile: fileLoc, config}) || spec;
     }
     const aliasEntry = findMatchingAliasEntry(config, spec);

--- a/snowpack/src/commands/build.ts
+++ b/snowpack/src/commands/build.ts
@@ -19,7 +19,7 @@ import {
   SnowpackBuildResult,
   SnowpackConfig,
 } from '../types';
-import {deleteFromBuildSafe, isRemoteUrl} from '../util';
+import {deleteFromBuildSafe, isPathImport, isRemoteUrl} from '../util';
 import {startServer} from './dev';
 
 function getIsHmrEnabled(config: SnowpackConfig) {
@@ -155,13 +155,17 @@ export async function build(commandOptions: CommandOptions): Promise<SnowpackBui
         logger.debug(`ADD: ${importedUrl}`);
         if (isRemoteUrl(importedUrl)) {
           // do nothing
-        } else if (importedUrl.startsWith('./') || importedUrl.startsWith('../')) {
-          logger.warn(`warn: import "${importedUrl}" of "${fileUrl}" could not be resolved.`);
-        } else if (!importedUrl.startsWith('/')) {
+        } else if (isPathImport(importedUrl)) {
+          if (importedUrl[0] === '/') {
+            if (!allFileUrlsUnique.has(importedUrl)) {
+              allFileUrlsUnique.add(importedUrl);
+              allFileUrlsToProcess.push(importedUrl);
+            }
+          } else {
+            logger.warn(`warn: import "${importedUrl}" of "${fileUrl}" could not be resolved.`);
+          }
+        } else {
           allBareModuleSpecifiers.push(installTarget);
-        } else if (!allFileUrlsUnique.has(importedUrl)) {
-          allFileUrlsUnique.add(importedUrl);
-          allFileUrlsToProcess.push(importedUrl);
         }
       }
     }

--- a/snowpack/src/config.ts
+++ b/snowpack/src/config.ts
@@ -24,6 +24,7 @@ import {
   NATIVE_REQUIRE,
   REQUIRE_OR_IMPORT,
   removeTrailingSlash,
+  isPathImport,
 } from './util';
 import type {Awaited} from './util';
 
@@ -618,13 +619,7 @@ function resolveRelativeConfigAlias(
   const cleanAliasConfig = {};
   for (const [target, replacement] of Object.entries(aliasConfig)) {
     const isDirectory = target.endsWith('/');
-    const isPath =
-      replacement === '.' ||
-      replacement === '..' ||
-      replacement.startsWith('./') ||
-      replacement.startsWith('../') ||
-      replacement.startsWith('/');
-    if (isPath) {
+    if (isPathImport(replacement)) {
       cleanAliasConfig[target] = isDirectory
         ? addTrailingSlash(path.resolve(configBase, replacement))
         : removeTrailingSlash(path.resolve(configBase, replacement));

--- a/snowpack/src/sources/local.ts
+++ b/snowpack/src/sources/local.ts
@@ -25,6 +25,7 @@ import {
   findMatchingAliasEntry,
   GLOBAL_CACHE_DIR,
   isJavaScript,
+  isPathImport,
   isRemoteUrl,
 } from '../util';
 import {installPackages} from './local-install';
@@ -124,7 +125,7 @@ export default {
       // relative files within the built package (ex: 'pkg/common/XXX-hash.js`).
       // We resolve these to a new kind of "internal" import URL that's different
       // from the normal, flattened URL for public imports.
-      if (spec.startsWith('./') || spec.startsWith('../')) {
+      if (isPathImport(spec)) {
         const newLoc = path.resolve(path.dirname(loc), spec);
         const resolvedSpec = slash(path.relative(installDest, newLoc));
         const publicImportEntry = Object.entries(packageImportMap.imports).find(
@@ -386,7 +387,7 @@ export default {
             if (isRemoteUrl(spec)) {
               continue;
             }
-            if (spec.startsWith('/') || spec.startsWith('./') || spec.startsWith('../')) {
+            if (isPathImport(spec)) {
               continue;
             }
             packageImports.add(spec);

--- a/snowpack/src/util.ts
+++ b/snowpack/src/util.ts
@@ -341,15 +341,7 @@ export function findMatchingAliasEntry(
   spec: string,
 ): {from: string; to: string; type: 'package' | 'path' | 'url'} | undefined {
   // Only match bare module specifiers. relative and absolute imports should not match
-  if (
-    spec === '.' ||
-    spec === '..' ||
-    spec.startsWith('./') ||
-    spec.startsWith('../') ||
-    spec.startsWith('/') ||
-    spec.startsWith('http://') ||
-    spec.startsWith('https://')
-  ) {
+  if (isPathImport(spec) || isRemoteUrl(spec)) {
     return undefined;
   }
 
@@ -386,6 +378,10 @@ export function getExtensionMatch(
   }
   // Return the first match, if one was found. Otherwise, return undefined.
   return extensionMatch ? [extensionPartial, extensionMatch] : undefined;
+}
+
+export function isPathImport(spec: string): boolean {
+  return spec[0] === '.' || spec[0] === '/';
 }
 
 export function isRemoteUrl(val: string): boolean {

--- a/test/build/config-alias/config-alias.test.js
+++ b/test/build/config-alias/config-alias.test.js
@@ -14,5 +14,6 @@ describe('config: alias', () => {
   it('generates imports as expected', () => {
     expect(files['/_dist_/index.html']).toMatchSnapshot();
     expect(files['/_dist_/index.js']).toMatchSnapshot();
+    expect(files['/_dist_/nested/foo.js']).toMatchSnapshot();
   });
 });

--- a/test/build/config-alias/src/nested/foo.js
+++ b/test/build/config-alias/src/nested/foo.js
@@ -1,0 +1,6 @@
+// importing index.js by shortcut
+import index from '.';
+import index_ from './';
+import index__ from '..';
+import index___ from '../';
+console.log(index, index_, index__, index___);

--- a/test/build/config-alias/src/nested/index.js
+++ b/test/build/config-alias/src/nested/index.js
@@ -1,0 +1,4 @@
+import sort from '../sort';
+console.log(sort);
+
+export default 'Button';

